qutebrowser/keyinput/modeman.py

```diff
@@ -294,6 +300,9 @@ def leave(self, mode, reason=None, maybe=False):
         self.clear_keychain()
         self.mode = usertypes.KeyMode.normal
         self.left.emit(mode, self.mode, self._win_id)
+        if mode in [usertypes.KeyMode.prompt, usertypes.KeyMode.yesno]:
+            self.enter(self._prev_mode,
+                       reason='restore mode before {}'.format(mode.name))
```

Comment:
This has the same bug your earlier PR had initially: It enters modes like the hint mode without having the necessary preparation.

Also, I really don't think it's worth the complexity: You can see for yourself how easy it is to introduce bugs, and I'm almost sure there are more corner cases when saving modes in different places. I also can't really see many cases where this would be useful (if I'm getting a quit confirmation, I wanted to quit anyways, so I don't care what mode I was in before).

Reply:
I see. The main reason for doing this is that tabs that when using tabs.mode_on_change=restore and passthrough mode is entered on ANY tabs, it will leave this mode whenever ANY prompt is shown (like for certificate/downloads, etc.).

What if I only save here for very specific modes and ONLY when tabs.mode_on_change=restore ?

Reply:
Probably okay to do it then.